pydantic/_internal/_dataclasses.py

```diff
@@ -92,10 +95,25 @@ def prepare_dataclass(
 
     core_config = generate_config(config, cls)
     cls.__pydantic_fields__ = fields
-    cls.__pydantic_validator__ = SchemaValidator(schema, core_config)
+    cls.__pydantic_validator__ = validator = SchemaValidator(schema, core_config)
     # this works because cls has been transformed into a dataclass by the time "cls" is called
     cls.__pydantic_serializer__ = SchemaSerializer(schema, core_config)
 
+    extra = config.get('extra', Extra.ignore)  # ignore is the default in V1
+    if extra is Extra.allow:
+        raise ValueError(
```

Comment:
Super minor, but should this be a TypeError? (I'm assuming we'll give a custom error type in the future anyway.)

If you want to leave as is and just get it merged I'm fine with that.

Reply:
No it should be a ValueError. It's the right type but the value isn't allowed.

Reply:
I meant it as a misconfiguration of the type you are creating, but your explanation makes sense.